control-plane-operator/controllers/hostedcontrolplane/cloud/powervs/powervs.go

```diff
@@ -0,0 +1,207 @@
+package powervs
+
+import (
+	"bytes"
+	"fmt"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	"strings"
+	"text/template"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	ccmContainerName       = "cloud-controller-manager"
+	kubeConfigFileBasePath = "/etc/kubernetes"
+	secretMountPath        = "/etc/vpc"
+	ccmConfigMapMountPath  = "/etc/ibm"
+)
+
+var ccmConfigTemplateData = `
+[global]
+version = 1.1.0
+[kubernetes]
+config-file = {{.ConfigFile}}
+[provider]
+cluster-default-provider = g2
+accountID = {{.AccountID}}
+clusterID = {{.ClusterID}}
+g2workerServiceAccountID = {{.G2workerServiceAccountID}}
+g2Credentials = {{.G2Credentials}}
+g2ResourceGroupName = {{.G2ResourceGroupName}}
+g2VpcSubnetNames = {{.G2VpcSubnetNames}}
+g2VpcName = {{.G2VpcName}}
+region = {{.Region}}
+powerVSCloudInstanceID = {{.PowerVSCloudInstanceID}}
+powerVSRegion = {{.PowerVSRegion}}
+powerVSZone = {{.PowerVSZone}}`
+
+var ccmConfigTemplate = template.Must(template.New("ccmConfigMap").Parse(ccmConfigTemplateData))
+
+func ReconcileCCMConfigMap(ccmConfig *corev1.ConfigMap, hcp *hyperv1.HostedControlPlane) (err error) {
+	config := map[string]string{
+		"ConfigFile":               fmt.Sprintf("%s/kubeconfig", kubeConfigFileBasePath),
+		"AccountID":                hcp.Spec.Platform.PowerVS.AccountID,
+		"ClusterID":                hcp.Name,
+		"G2workerServiceAccountID": hcp.Spec.Platform.PowerVS.AccountID,
+		"G2Credentials":            fmt.Sprintf("%s/%s", secretMountPath, "ibmcloud_api_key"),
+		"G2ResourceGroupName":      hcp.Spec.Platform.PowerVS.ResourceGroup,
+		"G2VpcSubnetNames":         hcp.Spec.Platform.PowerVS.VPC.Subnet,
+		"G2VpcName":                hcp.Spec.Platform.PowerVS.VPC.Name,
+		"Region":                   hcp.Spec.Platform.PowerVS.VPC.Region,
+		"PowerVSCloudInstanceID":   hcp.Spec.Platform.PowerVS.ServiceInstanceID,
+		"PowerVSRegion":            hcp.Spec.Platform.PowerVS.Region,
+		"PowerVSZone":              hcp.Spec.Platform.PowerVS.Zone,
+	}
+
+	configData := &bytes.Buffer{}
+	ccmConfigTemplate.Execute(configData, config)
+
+	if ccmConfig.Data == nil {
+		ccmConfig.Data = map[string]string{}
+	}
+
+	ccmConfig.Data[ccmConfig.Name] = configData.String()
+
+	return
+}
+
+func ReconcileCCMDeployment(deployment *appsv1.Deployment, hcp *hyperv1.HostedControlPlane, ccmConfig *corev1.ConfigMap, releaseImage *releaseinfo.ReleaseImage) (err error) {
```

Comment:
Have you considered this vs a container of KAS deployment? @JoelSpeed was there any consensus about that [here](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/out-of-tree-provider-support.md)?

@dharaneeshvrd Is there any recomendation/docs we can link for this provider? Is there any plan for OCP we can see to validate we are consistent here?

Reply:
In OCP, we decided to keep the CCM separate for maintenance sake. As a separate team owns the KCM/KAS vs CCM, it is easier internally to have a separation of concerns across team boundaries.

If within HyperShift, you prefer to run the CCM in a single deployment with other resources, that's up to you. Though, cons of this come when you start talking about RBAC. If you run it within a single deployment, you may in the future run into issues where you add large permission sets to the entire deployment, vs, having smaller, more well defined deployments where permissions can be stricter and more granularly managed

Reply:
This is the asset from CCM operator - https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/master/pkg/cloud/ibm/assets/deployment.yaml can be referred.

Reply:
@enxebre can you please suggest how we can handle the CCM deployment here? 
IMHO as per @JoelSpeed suggestion If we keep CCM deployment separately, we can handle the RBAC more granularly in future.

Reply:
If the rbac applies to deployment A or B I think is cosmetic in this case, so I'd consider performance and practical implications as the driving factor, I thought ibm cloud use to run it as sidecar in other topologies so curious about that experience. Let's keep it as a separate deployment for simplicity of the code base for now.